File: C++/HARD/2463. Minimum Total Distance Traveled.cpp
#define veryLarge 2147483647000
class Solution
{
    vector<vector<long long>> dp;

public:
    long long minimumTotalDistance(vector<int> &robot, vector<vector<int>> &factory)
    {
        sort(robot.begin(), robot.end());
        sort(factory.begin(), factory.end());
        int rSize = robot.size();
        int fSize = factory.size();
        dp = vector<vector<long long>>(fSize + 1, vector<long long>(rSize + 1, veryLarge));
        long long answer = veryLarge;
        dp[0][0] = 0;
        // 팩토리
        for (int i = 0; i < fSize; i++)
        {
            // 로봇
            for (int j = 0; j < rSize; j++)
            {
                if (dp[i][j] == veryLarge)
                    continue;
                dp[i + 1][j] = min(dp[i][j], dp[i + 1][j]);
                long long ds = 0;
                for (int k = j + 1; k <= min(rSize, j + factory[i][1]); k++)
                {
                    // cout << k << dp[i][j] + ds;
                    ds += abs(factory[i][0] - robot[k - 1]);
                    dp[i + 1][k] = min(dp[i + 1][k], dp[i][j] + ds);
                }
            }
        }

        for (int i = 0; i <= fSize; i++)
        {
            if (dp[i][rSize] != -1)
            {
                answer = min(answer, dp[i][rSize]);
            }
        }
        return answer;
    }
};

File: 2463-minimum-total-distance-traveled/2463-minimum-total-distance-traveled.java
class Solution {
    Long MAXIMUM = (long) 1e13;



    public long minimumTotalDistance(List<Integer> robot, int[][] factory) {
        Collections.sort(robot);
        // List<int[]> newFactory = new ArrayList<>(); 
        List<Integer> newNewFactory = new ArrayList<>(); 
        for (int[] fac : factory)
        {
            for (int j = 0; j < fac[1]; j++)
            {
                newNewFactory.add(fac[0]);
            }
        }

        Collections.sort(newNewFactory);
        
        int fn = newNewFactory.size();
        int n = robot.size();

        long[][] DP = new long[n + 1][fn + 1];


        for (int i = 0; i <= n; i++)
        {
            for (int j = 0; j <= fn; j++)
            {
                DP[i][j] = MAXIMUM;
            }
        }
        for (int j = 0; j < fn; j++)
        {
            DP[0][j] = 0;
        }

        for (int i = 0; i < n; i++)
        {
            // 아님 여기서 prefixSum을 쓰나? 근데 그것도 이상한데 
            for (int j = 0; j < fn; j++)
            {
                // 너가 얘좀 처리할래? 하고 미룬상태임
                DP[i][j + 1] = Math.min(
                    DP[i][j + 1], DP[i][j]
                );

                // 야 내가 처리할게 로 하고 
                DP[i + 1][j + 1] = Math.min(
                    DP[i + 1][j + 1],
                    DP[i][j] + Math.abs(robot.get(i) - newNewFactory.get(j))
                );

            }
        }


        long answer = MAXIMUM;
        for (int j = 0; j <= fn; j++)
        {
            answer = Math.min(answer, DP[n][j]);
        }
        return answer;
    }
}